### PR TITLE
[BEAM-6098] Support lookup join symmetric in left/right inputs

### PR DESCRIPTION
**Pull request by kennknowles**
A minor refactor and also support lookup joins on either left or right.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by kennknowles** [comment]
Some corner cases in various methods still being worked out, but FYI on the change.

**Comment by kennknowles** [comment]
OK, I got the tests fixed up and passing. I admit that it could be prettier.

**Comment by kennknowles** [comment]
Rebased on `master` since it seemed to be hitting a failure on `master`.

**Comment by kennknowles** [comment]
R: @akedin

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamJoinRel.java [hunk 1]
```diff
@@ -120,35 +120,91 @@ public Join copy(
...
-    return new Transform();
+    if (isSideInputLookupJoin()) {
+      return new SideInputLookupJoin();
+    } else {
+      return new Transform();
+    }
+  }
+
+  private boolean isSideInputLookupJoin() {
+    return seekableInputIndex().isPresent() && nonSeekableInputIndex().isPresent();
```
[the remark this answers is not in the document]
Comment (kennknowles) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamJoinRel.java]: It is a good question. I was just copying existing logic for symmetry. If they are both seekable it seems still fine to do this sort of join. You would want to make the main input the smaller one. But we don't have sizes yet anyhow.

[the remark this answers is not in the document]
Comment (XuMingmin) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamJoinRel.java]: If both are seekable, I would expect a failure thrown.

**Pull request merged**